### ISSUE TITLE
Improper use of closures not detected

### ISSUE THREAD

**Issue by remexre** [rust-lang/rust#51486]
The following code gets "overflow evaluating the requirement `[closure@src/main.rs:20:19: 20:28]: std::ops::FnMut<(i32,)>`" if the line in `main` is uncommented.
If the error should exist (which I'm unsure about), it should be detected without needing to call `map`.
If not, that's a bug.

```rust
enum Foo<T> {
    A(T),
    B(Box<Foo<T>>, Box<Foo<T>>),
}

impl<T> Foo<T> {
    fn map<F, T2>(self, mut f: F) -> Foo<T2>
    where
        F: FnMut(T) -> T2,
    {
        match self {
            Foo::A(x) => Foo::A(f(x)),
            Foo::B(x, y) => Foo::B(Box::new(x.map(&mut f)), Box::new(y.map(&mut f))),
        }
    }
}

fn main() {
    // Foo::A(0).map(|x| x + 1);
}
```

**Comment by estebank** [comment]
Closing as duplicate of #50043.

**Comment by estebank** [comment]
Ah, realized that the report is quite more specific than the linked ticket. Keeping open.
